### ISSUE TITLE
为什么识别时间比较长？

### ISSUE THREAD

**Issue by apachexc** [pengzhendong/streaming-sensevoice#60]
(sense3.8env) [root@iZf8z0twy80rkax8w9mpopZ streaming-sensevoice]# python ./offline_main.py  --a ./data/test_16k.wav --c 停止
初始化模型...
2025-11-28 16:16:34,823 - modelscope - INFO - PyTorch version 2.3.0 Found.
2025-11-28 16:16:34,825 - modelscope - INFO - Loading ast index from /root/.cache/modelscope/ast_indexer
2025-11-28 16:16:35,074 - modelscope - INFO - Loading done! Current index file version is 1.9.5, with md5 37a6f0274d570560397ff3d0c08b4103 and a total number of 945 components indexed
Loading remote code failed: model, No module named 'model'
模型预热中...
识别音频：./data/test_16k.wav...

===== CPU优化识别结果 =====
识别文本：甚至出现交易几乎停止的情况
推理耗时：5294.07 毫秒
音频时长：4.20 秒
CPU线程数：4
================================
(sense3.8env) [root@iZf8z0twy80rkax8w9mpopZ streaming-sensevoice]# lscpu | grep "CPU(s):"
CPU(s):              4
NUMA node0 CPU(s):   0-3

**Comment by apachexc** [comment]
**测试代码：**


import soundfile as sf
import argparse
import time
import numpy as np
import torch
from streaming_sensevoice import StreamingSenseVoice


def main():
    # 彻底屏蔽所有日志（包括modelscope）
    import logging
    logging.getLogger().setLevel(logging.CRITICAL + 1)

    # 解析参数
    parser = argparse.ArgumentParser(description="SenseVoice CPU离线稳定优化版")
    parser.add_argument("-a", "--audio", required=True, help="语音文件路径")
    parser.add_argument("-c", "--contexts", nargs="+", default=["停止"], help="自定义热词")
    args = parser.parse_args()

    # ========== CPU多核性能拉满（核心优化） ==========
    # 根据服务器CPU核心数调整（例如32核设为32，8核设为8）
    cpu_core_num = 16  # 请根据你的服务器实际核心数修改
    torch.set_num_threads(cpu_core_num)
    torch.set_num_interop_threads(cpu_core_num)

    # ========== 初始化模型（仅用原生支持的参数） ==========
    print("初始化模型...")
    model = StreamingSenseVoice(contexts=args.contexts)  # 原生接口，无额外参数

    # ========== 模型预热（消除首次加载开销，关键） ==========
    print("模型预热中...")
    warmup_audio = [0] * 16000  # 1秒空音频（匹配模型输入格式）
    model.streaming_inference(warmup_audio, is_last=True)  # 仅调用原生流式接口

    # ========== 音频预处理（极简高效） ==========
    samples, sr = sf.read(args.audio, dtype='int16')  # 直接读取int16，避免格式转换
    # 转单声道（若需要）
    if len(samples.shape) > 1:
        samples = samples.mean(axis=1).astype(np.int16)
    # 强制16k采样率（模型要求）
    if sr != 16000:
        import torchaudio
        resampler = torchaudio.transforms.Resample(sr, 16000, dtype=torch.int16)
        samples = resampler(torch.from_numpy(samples)).numpy()
    samples = samples.tolist()

    # ========== 实际识别（仅计时推理过程） ==========
    print(f"识别音频：{args.audio}...")
    start_time = time.time()
    final_result = None
    # 一次性传入完整音频+标记is_last=True（非流式极速模式）
    for res in model.streaming_inference(samples, is_last=True):
        final_result = res
    end_time = time.time()

    # ========== 输出结果 ==========
    elapsed_time = (end_time - start_time) * 1000
    audio_duration = len(samples) / 16000
    print("\n===== CPU优化识别结果 =====")
    print(f"识别文本：{final_result['text']}")
    print(f"推理耗时：{elapsed_time:.2f} 毫秒")
    print(f"音频时长：{audio_duration:.2f} 秒")
    print(f"CPU线程数：{torch.get_num_threads()}")
    print("================================")


if __name__ == "__main__":
    main()

**Comment by pengzhendong** [comment]
几个原因叠加：

1. **线程数设错了**：你是 4 核机器但设了 `torch.set_num_threads(16)`，线程超售反而更慢，改成物理核数 4。
2. **预热其实没生效**：`streaming_inference` 是生成器，直接调用不迭代的话函数体一行都不会执行。要 `list(model.streaming_inference(...))` 或 for 循环消费掉才算真的跑了。
3. **流式接口本身是重复计算的**：当前实现每个 chunk 都会把积累的全部历史特征重新过一遍 encoder（全历史双向注意力，效果最好但计算量随音频长度平方增长）。如果只是离线转写单个文件，一次性喂完 + `is_last=True` 只跑一遍即可；长音频流式场景可以用 `max_history` 参数限制历史帧数来封顶计算量。
4. 纯 CPU 跑 SenseVoiceSmall（约 2.3 亿参数）本来就不快，4 核 RTF 在 1 附近是正常水平；要快就上 GPU 或者换 ONNX 量化方案。

另外日志里 `Loading remote code failed: model, No module named 'model'` 是 funasr/modelscope 的无害告警，不影响推理。
